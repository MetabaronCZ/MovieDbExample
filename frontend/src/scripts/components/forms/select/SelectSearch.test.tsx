import { act } from '@testing-library/react';
import { describe, expect, it, jest } from '@jest/globals';
import {
  fireEvent,
  getAllByRole,
  getByRole,
  getByTestId,
  getByText,
  getByTitle,
  queryByTestId,
} from '@testing-library/dom';

import { SelectSearch } from 'components/forms/select/SelectSearch';
import { TestComponentWrapper, waitForComponent } from 'test-utils/component';

// mock Loader component
jest.mock('../../common/Loader.tsx', () => ({
  Loader: () => <div data-testid="loader-mock" />,
}));

describe('components/forms/select/SelectSearch', () => {
  it('should render properly', async () => {
    const { container } = await waitForComponent(
      <SelectSearch
        id="ID"
        value={[{ title: 'Item 1', value: 1 }]}
        options={[
          { title: 'Item 1', value: 1 },
          { title: 'Item 2', value: 2 },
        ]}
        onSelect={() => null}
        onSearch={() => null}
      />,
      { wrapper: TestComponentWrapper },
    );
    expect(container).toMatchSnapshot();

    const list = getByRole(container, 'list');
    expect(list).toBeInTheDocument();

    const listItems = getAllByRole(list, 'listitem');
    expect(listItems.length).toEqual(2);

    // selected items
    const selectedItemButton = getByText(listItems[0], 'Item 1');
    expect(selectedItemButton).toBeInTheDocument();
    expect(selectedItemButton.nodeName.toLowerCase()).toEqual('button');

    const selectedItemClearButton = getByTitle(listItems[0], 'Odebrat');
    expect(selectedItemClearButton).toBeInTheDocument();
    expect(selectedItemClearButton.nodeName.toLowerCase()).toEqual('button');

    // search input
    const search = getByRole(listItems[1], 'textbox');
    expect(search).toBeInTheDocument();
    expect(search).toHaveAttribute('id', 'ID');
    expect(search).toHaveAttribute('value', '');
    expect(search).toHaveAttribute('placeholder', 'Hledat');

    // clear selection button
    const clearButton = list.nextSibling;
    expect(clearButton).toBeInTheDocument();
    expect(clearButton?.nodeName.toLowerCase()).toEqual('button');
    expect(clearButton).toHaveAttribute('title', 'ZruÅ¡it');
  });

  it('should render empty-value variant', async () => {
    const { container } = await waitForComponent(
      <SelectSearch
        value={[]}
        options={[
          { title: 'Item 1', value: 1 },
          { title: 'Item 2', value: 2 },
        ]}
        onSelect={() => null}
        onSearch={() => null}
      />,
      { wrapper: TestComponentWrapper },
    );
    expect(container).toMatchSnapshot();

    const list = getByRole(container, 'list');
    expect(list).toBeInTheDocument();

    const listItems = getAllByRole(list, 'listitem');
    expect(listItems.length).toEqual(1);

    // search input
    const search = getByRole(listItems[0], 'textbox');
    expect(search).toBeInTheDocument();

    // no selection >> no clear button
    const clearButton = list.nextSibling;
    expect(clearButton).not.toBeInTheDocument();
  });

  it('should render empty-options variant', async () => {
    const { container } = await waitForComponent(
      <SelectSearch
        value={[]}
        options={[]}
        onSelect={() => null}
        onSearch={() => null}
      />,
      { wrapper: TestComponentWrapper },
    );
    expect(container).toMatchSnapshot();

    const list = getByRole(container, 'list');
    expect(list).toBeInTheDocument();

    const listItems = getAllByRole(list, 'listitem');
    expect(listItems.length).toEqual(1);

    // search input
    const search = getByRole(listItems[0], 'textbox');
    expect(search).toBeInTheDocument();

    // no selection >> no clear button
    const clearButton = list.nextSibling;
    expect(clearButton).not.toBeInTheDocument();
  });

  it('should clear selected item', async () => {
    const onSelect = jest.fn();

    const { container } = await waitForComponent(
      <SelectSearch
        value={[
          { title: 'Item 1', value: 1 },
          { title: 'Item 2', value: 2 },
        ]}
        options={[
          { title: 'Item 1', value: 1 },
          { title: 'Item 2', value: 2 },
        ]}
        onSelect={onSelect}
        onSearch={() => null}
      />,
      { wrapper: TestComponentWrapper },
    );
    expect(onSelect).not.toHaveBeenCalled();

    const list = getByRole(container, 'list');
    expect(list).toBeInTheDocument();

    const listItems = getAllByRole(list, 'listitem');
    expect(listItems.length).toEqual(3); // 2 items + search input

    // click Item 1 clear button
    const item1Clear = listItems[0].querySelectorAll('button')[1];

    act(() => {
      fireEvent.click(item1Clear);
    });
    expect(onSelect).toHaveBeenCalledTimes(1);
    expect(onSelect).toHaveBeenCalledWith([2]); // array of item IDs left

    // click Item 2 clear button
    const item2Clear = listItems[1].querySelectorAll('button')[1];

    act(() => {
      fireEvent.click(item2Clear);
    });
    expect(onSelect).toHaveBeenCalledTimes(2);
    expect(onSelect).toHaveBeenCalledWith([1]); // array of item IDs left
  });

  it('should not clear selected item if disabled', async () => {
    const onSelect = jest.fn();

    const { container } = await waitForComponent(
      <SelectSearch
        value={[{ title: 'Item 1', value: 1, disabled: true }]}
        options={[{ title: 'Item 1', value: 1 }]}
        onSelect={onSelect}
        onSearch={() => null}
      />,
      { wrapper: TestComponentWrapper },
    );
    expect(onSelect).not.toHaveBeenCalled();

    const list = getByRole(container, 'list');
    expect(list).toBeInTheDocument();

    const listItems = getAllByRole(list, 'listitem');
    expect(listItems.length).toEqual(2); // selected item + search input

    // click Item 1 clear button
    const item1Clear = listItems[0].querySelectorAll('button')[1];

    act(() => {
      fireEvent.click(item1Clear);
    });
    expect(onSelect).not.toHaveBeenCalled();
  });

  it('should clear all selected items', async () => {
    const onSelect = jest.fn();

    const { container } = await waitForComponent(
      <SelectSearch
        value={[
          { title: 'Item 1', value: 1 },
          { title: 'Item 2', value: 2 },
        ]}
        options={[
          { title: 'Item 1', value: 1 },
          { title: 'Item 2', value: 2 },
        ]}
        onSelect={onSelect}
        onSearch={() => null}
      />,
      { wrapper: TestComponentWrapper },
    );
    expect(onSelect).not.toHaveBeenCalled();

    const list = getByRole(container, 'list');
    expect(list).toBeInTheDocument();

    const listItems = getAllByRole(list, 'listitem');
    expect(listItems.length).toEqual(3); // 2 items + search input

    // clear all selected button
    const clearButton = list.nextSibling;

    if (!clearButton) {
      throw new Error('Clear button not found!');
    }
    act(() => {
      fireEvent.click(clearButton);
    });
    expect(onSelect).toHaveBeenCalledTimes(1);
    expect(onSelect).toHaveBeenCalledWith([]);
  });

  it('should not clear selected items when disabled', async () => {
    const onSelect = jest.fn();

    const { container } = await waitForComponent(
      <SelectSearch
        value={[
          { title: 'Item 1', value: 1 },
          { title: 'Item 2', value: 2 },
        ]}
        options={[
          { title: 'Item 1', value: 1 },
          { title: 'Item 2', value: 2 },
        ]}
        disabled
        onSelect={onSelect}
        onSearch={() => null}
      />,
      { wrapper: TestComponentWrapper },
    );
    expect(onSelect).not.toHaveBeenCalled();

    const list = getByRole(container, 'list');
    expect(list).toBeInTheDocument();

    const listItems = getAllByRole(list, 'listitem');
    expect(listItems.length).toEqual(3); // 2 items + search input

    // clear all selected button
    const clearButton = list.nextSibling;

    if (!clearButton) {
      throw new Error('Clear button not found!');
    }
    act(() => {
      fireEvent.click(clearButton);
    });
    expect(onSelect).not.toHaveBeenCalled();
  });

  it('should be searchable', async () => {
    jest.useFakeTimers();

    const onSearch = jest.fn();
    const onSelect = jest.fn();

    const { container } = await waitForComponent(
      <SelectSearch
        value={[]}
        options={[
          { title: 'Item 1', value: 1 },
          { title: 'Item 2', value: 2 },
        ]}
        onSelect={onSelect}
        onSearch={onSearch}
      />,
      { wrapper: TestComponentWrapper },
    );
    expect(onSelect).not.toHaveBeenCalled();
    expect(onSearch).not.toHaveBeenCalled();

    let lists = getAllByRole(container, 'list');
    expect(lists.length).toEqual(1);

    const listItems = getAllByRole(container, 'listitem');
    expect(listItems.length).toEqual(1); // search input only

    const search = getByRole(listItems[0], 'textbox');
    expect(search).toBeInTheDocument();

    // search query string
    act(() => {
      fireEvent.change(search, { target: { value: 'QUERY' } });
      jest.runAllTimers();
    });

    expect(onSearch).toHaveBeenCalledTimes(1);
    expect(onSearch).toHaveBeenCalledWith('QUERY');

    // result list should be opened
    lists = getAllByRole(container, 'list');
    expect(lists.length).toEqual(2);

    const results = getAllByRole(lists[1], 'listitem');
    expect(results.length).toEqual(2);
    expect(results[0]).toHaveTextContent('Item 1');
    expect(results[1]).toHaveTextContent('Item 2');

    // select Item 2
    const item2Button = getByRole(results[1], 'button');

    act(() => {
      fireEvent.click(item2Button);
    });
    expect(onSelect).toHaveBeenCalledTimes(1);
    expect(onSelect).toHaveBeenCalledWith([2]);

    // results should be still opened for further selection
    lists = getAllByRole(container, 'list');
    expect(lists.length).toEqual(2);

    // select Item 1
    const item1Button = getByRole(results[0], 'button');

    act(() => {
      fireEvent.click(item1Button);
    });
    expect(onSelect).toHaveBeenCalledTimes(2);
    expect(onSelect).toHaveBeenCalledWith([1]);

    // close result by clicking Escape key
    act(() => {
      fireEvent.keyUp(document, { key: 'Escape' });
    });
    lists = getAllByRole(container, 'list');
    expect(lists.length).toEqual(1);

    jest.useRealTimers();
  });

  it('should not search / open when disabled', async () => {
    jest.useFakeTimers();

    const onSearch = jest.fn();
    const onSelect = jest.fn();

    const { container } = await waitForComponent(
      <SelectSearch
        value={[]}
        options={[
          { title: 'Item 1', value: 1 },
          { title: 'Item 2', value: 2 },
        ]}
        disabled
        onSelect={onSelect}
        onSearch={onSearch}
      />,
      { wrapper: TestComponentWrapper },
    );
    expect(onSelect).not.toHaveBeenCalled();
    expect(onSearch).not.toHaveBeenCalled();

    let lists = getAllByRole(container, 'list');
    expect(lists.length).toEqual(1);

    const listItems = getAllByRole(container, 'listitem');
    expect(listItems.length).toEqual(1); // search input only

    const search = getByRole(listItems[0], 'textbox');
    expect(search).toBeInTheDocument();
    expect(search).toHaveAttribute('disabled');

    // try search query string
    act(() => {
      fireEvent.change(search, { target: { value: 'QUERY' } });
      jest.runAllTimers();
    });
    expect(onSearch).not.toHaveBeenCalled();

    // result list should not be opened
    lists = getAllByRole(container, 'list');
    expect(lists.length).toEqual(1);
  });

  it('should reflect loading state', async () => {
    jest.useFakeTimers();

    const { container } = await waitForComponent(
      <SelectSearch
        value={[]}
        options={[
          { title: 'Item 1', value: 1 },
          { title: 'Item 2', value: 2 },
        ]}
        loading
        onSelect={() => null}
        onSearch={() => null}
      />,
      { wrapper: TestComponentWrapper },
    );

    // no Loader / results closed
    let loader = queryByTestId(container, 'loader-mock');
    expect(loader).not.toBeInTheDocument();

    // search input
    const search = getByRole(container, 'textbox');
    expect(search).toBeInTheDocument();

    // search query string
    act(() => {
      fireEvent.change(search, { target: { value: 'QUERY' } });
      jest.runAllTimers();
    });

    // Loader present
    loader = getByTestId(container, 'loader-mock');
    expect(loader).toBeInTheDocument();

    jest.useRealTimers();
  });
});
