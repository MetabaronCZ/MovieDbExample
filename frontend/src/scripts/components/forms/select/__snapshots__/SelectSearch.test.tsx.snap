// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`components/forms/select/SelectSearch should render empty-options variant 1`] = `
.c5 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  display: block;
  border: none;
  background: transparent;
  outline: none;
}

.c5::placeholder {
  font-style: italic;
  color: inherit;
  opacity: 0.5;
}

.c5:hover:not(:disabled) {
  background-color: #f2f7fa;
}

.c5:disabled {
  color: #98c3d9;
}

.c2 {
  list-style-type: none;
  flex: 1;
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  align-items: center;
  gap: 4px;
  padding: 4px 0;
}

.c3 {
  display: flex;
  flex-direction: row;
  align-items: center;
  gap: 1px;
}

.c4 {
  flex: 1;
  padding-left: 4px;
}

.c6 {
  width: 100%;
}

.c0 {
  position: relative;
  display: inline-block;
  width: 100%;
}

.c1 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  display: inline-flex;
  flex-direction: row;
  align-items: center;
  gap: 12px;
  line-height: 1;
  font-weight: 700;
  height: 40px;
  padding: 0 16px;
  padding-left: 12px;
  padding-right: 12px;
  border: 2px solid #002d56;
  background: #fff;
  width: 100%;
  height: 40px;
  gap: 8px;
  padding-left: 12px;
  padding-right: 12px;
  height: auto;
  gap: 8px;
  min-height: 40px;
  padding-right: 4px;
}

.c1:hover:not(:disabled) {
  background-color: #f2f7fa;
}

.c1:focus {
  outline: 1px dashed #002d56;
}

.c1:disabled {
  color: #98c3d9;
  border-color: #98c3d9;
  background-color: #fff;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <ul
        class="c2"
      >
        <li
          class="c3 c4"
        >
          <input
            autocomplete="one-time-code"
            class="c5 c6"
            placeholder="Hledat"
            type="text"
            value=""
          />
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`components/forms/select/SelectSearch should render empty-value variant 1`] = `
.c5 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  display: block;
  border: none;
  background: transparent;
  outline: none;
}

.c5::placeholder {
  font-style: italic;
  color: inherit;
  opacity: 0.5;
}

.c5:hover:not(:disabled) {
  background-color: #f2f7fa;
}

.c5:disabled {
  color: #98c3d9;
}

.c2 {
  list-style-type: none;
  flex: 1;
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  align-items: center;
  gap: 4px;
  padding: 4px 0;
}

.c3 {
  display: flex;
  flex-direction: row;
  align-items: center;
  gap: 1px;
}

.c4 {
  flex: 1;
  padding-left: 4px;
}

.c6 {
  width: 100%;
}

.c0 {
  position: relative;
  display: inline-block;
  width: 100%;
}

.c1 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  display: inline-flex;
  flex-direction: row;
  align-items: center;
  gap: 12px;
  line-height: 1;
  font-weight: 700;
  height: 40px;
  padding: 0 16px;
  padding-left: 12px;
  padding-right: 12px;
  border: 2px solid #002d56;
  background: #fff;
  width: 100%;
  height: 40px;
  gap: 8px;
  padding-left: 12px;
  padding-right: 12px;
  height: auto;
  gap: 8px;
  min-height: 40px;
  padding-right: 4px;
}

.c1:hover:not(:disabled) {
  background-color: #f2f7fa;
}

.c1:focus {
  outline: 1px dashed #002d56;
}

.c1:disabled {
  color: #98c3d9;
  border-color: #98c3d9;
  background-color: #fff;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <ul
        class="c2"
      >
        <li
          class="c3 c4"
        >
          <input
            autocomplete="one-time-code"
            class="c5 c6"
            placeholder="Hledat"
            type="text"
            value=""
          />
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`components/forms/select/SelectSearch should render properly 1`] = `
.c4 {
  display: block;
  border: none;
  background: transparent;
  outline: none;
  cursor: pointer;
}

.c4:disabled {
  cursor: not-allowed;
}

.c10 {
  display: inline-flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;
  width: 32px;
  height: 32px;
}

.c10:hover:not(:disabled) {
  background-color: #f2f7fa;
}

.c10:focus {
  outline: 1px dashed #002d56;
}

.c8 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  display: block;
  border: none;
  background: transparent;
  outline: none;
}

.c8::placeholder {
  font-style: italic;
  color: inherit;
  opacity: 0.5;
}

.c8:hover:not(:disabled) {
  background-color: #f2f7fa;
}

.c8:disabled {
  color: #98c3d9;
}

.c2 {
  list-style-type: none;
  flex: 1;
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  align-items: center;
  gap: 4px;
  padding: 4px 0;
}

.c3 {
  display: flex;
  flex-direction: row;
  align-items: center;
  gap: 1px;
}

.c7 {
  flex: 1;
  padding-left: 4px;
}

.c9 {
  width: 100%;
}

.c5 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  font-size: 12px;
  line-height: 24px;
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;
  padding: 0 8px;
  background: #98c3d9;
  white-space: nowrap;
}

.c5:hover:not(:disabled) {
  background-color: #0af;
}

.c5:focus {
  outline: 1px dashed #002d56;
}

.c5:disabled {
  cursor: initial;
}

.c6 {
  width: 24px;
  height: 24px;
  padding: 0;
}

.c0 {
  position: relative;
  display: inline-block;
  width: 100%;
}

.c1 {
  font-family: Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 24px;
  color: #002d56;
  display: inline-flex;
  flex-direction: row;
  align-items: center;
  gap: 12px;
  line-height: 1;
  font-weight: 700;
  height: 40px;
  padding: 0 16px;
  padding-left: 12px;
  padding-right: 12px;
  border: 2px solid #002d56;
  background: #fff;
  width: 100%;
  height: 40px;
  gap: 8px;
  padding-left: 12px;
  padding-right: 12px;
  height: auto;
  gap: 8px;
  min-height: 40px;
  padding-right: 4px;
}

.c1:hover:not(:disabled) {
  background-color: #f2f7fa;
}

.c1:focus {
  outline: 1px dashed #002d56;
}

.c1:disabled {
  color: #98c3d9;
  border-color: #98c3d9;
  background-color: #fff;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <ul
        class="c2"
      >
        <li
          class="c3"
        >
          <button
            class="c4 c5"
            type="button"
          >
            Item 1
          </button>
          <button
            class="c4 c5 c6"
            title="Odebrat"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-xmark fa-sm "
              data-icon="xmark"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 384 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M342.6 150.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 210.7 86.6 105.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L146.7 256 41.4 361.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L192 301.3 297.4 406.6c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L237.3 256 342.6 150.6z"
                fill="currentColor"
              />
            </svg>
          </button>
        </li>
        <li
          class="c3 c7"
        >
          <input
            autocomplete="one-time-code"
            class="c8 c9"
            id="ID"
            placeholder="Hledat"
            type="text"
            value=""
          />
        </li>
      </ul>
      <button
        class="c4 c10"
        title="ZruÅ¡it"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-xmark fa-lg "
          color="#002d56"
          data-icon="xmark"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 384 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M342.6 150.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 210.7 86.6 105.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L146.7 256 41.4 361.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L192 301.3 297.4 406.6c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L237.3 256 342.6 150.6z"
            fill="currentColor"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
